eval/tricks.rkt

```diff
@@ -0,0 +1,301 @@
+       [(bfinfinite? hi) (- (min (mpfr-exp lo) 0) (get-slack iter))]
+       [(bfinfinite? lo) (- (min (mpfr-exp hi) 0) (get-slack iter))]
+       [else (- (min (mpfr-exp lo) (mpfr-exp hi) 0) (get-slack iter))])]
+    [else
+     (cond
+       ; Can't both be inf, since:
+       ;  - [inf, inf] not a valid interval
+       ;  - [-inf, inf] crosses zero
+       [(bfinfinite? lo) (mpfr-exp hi)]
+       [(bfinfinite? hi) (mpfr-exp lo)]
+       [else (min (mpfr-exp lo) (mpfr-exp hi))])]))
+
+(define (logspan x)
+  #;(define lo (ival-lo x))
+  #;(define hi (ival-hi x))
+  #;(if (or (bfzero? lo) (bfinfinite? lo) (bfzero? hi) (bfinfinite? hi))
+        (get-slack)
+        (+ (abs (- (mpfr-exp lo) (mpfr-exp hi))) 1))
+  0)
+
+; Function calculates an ampl factor per input for a certain output and inputs using condition formulas,
+;   where an ampl is an additional precision that needs to be added to srcs evaluation so,
+;   that the output will be fixed in its precision when evaluating again
+; Additionaly, the function retures lower bound of ampl factor for the early exit mechanism
+; Output: '( '(upper-ampl-bound lower-ampl-bound) ...) with len(srcs) number of elements
+(define (get-bounds op z srcs)
+  (case (object-name op)
+    [(ival-mult)
+     ; k = 1: logspan(y)
+     ; k = 2: logspan(x)
+     (define x (first srcs))
+     (define y (second srcs))
+     (list (list (logspan y) 0) ; bounds per x
+           (list (logspan x) 0))] ; bounds per y
+
+    [(ival-div)
+     ; k = 1: logspan(y)
+     ; k = 2: logspan(x) + 2 * logspan(y)
+     (define x (first srcs))
+     (define y (second srcs))
+     (list (list (logspan y) 0) ; bounds per x
+           (list (+ (logspan x) (* 2 (logspan y))) 0))] ; bounds per y
+
+    [(ival-sqrt ival-cbrt)
+     ; sqrt: logspan(x)/2 - 1
+     ; cbrt: logspan(x)*2/3 - 1
+     (define x (first srcs))
+     (list (list (quotient (logspan x) 2) 0))]
+
+    [(ival-add ival-sub)
+     ; k = 1: maxlog(x) - minlog(z)
+     ; k = 2: maxlog(y) - minlog(z)
+     (define x (first srcs))
+     (define y (second srcs))
+
+     (list (list (- (maxlog x) (minlog z))
+                 (- (minlog x #:underestimate #t) (maxlog z #:underestimate #t))) ; bounds per x
+           (list (- (maxlog y) (minlog z))
+                 (- (minlog y #:underestimate #t) (maxlog z #:underestimate #t))))] ; bounds per y
+
+    [(ival-pow)
+     ; k = 1: maxlog(y) + logspan(x) + logspan(z)
+     ; k = 2: maxlog(y) + max(|minlog(x)|,|maxlog(x)|) + logspan(z)
+     (define x (first srcs))
+     (define y (second srcs))
+
+     ; when output crosses zero and x is negative - means that y was fractional and not fixed (specific of Rival)
+     ; solution - add more slack for y to converge
+     (define y-slack
+       (if (and (crosses-zero? z) (bfnegative? (ival-lo x)))
+           (get-slack)
+           0))
+
+     ; when output is (ival 0.bf 1.bf) - it means that x was close to 1 or 0 but not narrow enough
+     (define x-slack
```

Comment:
This part of the formulas has been updated. 
I previously thought that it does not matter, however, turned out that for Harley example this part helps to sample 5 rare edge cases